File: tools/tabula-distro/src/tabula_distro/semver.py
"""Minimal SemVer constraint parser (no third-party deps).

Supports a tiny subset sufficient for our distro/bundle compatibility checks:

  >=1.2.3   <2.0.0   >1.0   <=1.5.0   ==1.4.2   1.4.2

Constraints are comma-separated and joined with AND:

  ">=0.8.0,<1.0.0"

Versions must be strict ``MAJOR.MINOR.PATCH`` integers, optionally followed by
a ``-pre.release`` suffix (which we tolerate but compare lexically — good enough
for early development; we don't ship pre-releases yet).
"""
from __future__ import annotations

import re
from dataclasses import dataclass


_VERSION_RE = re.compile(r"^(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?$")
_OPS = ("<=", ">=", "==", "<", ">", "=")


class VersionError(ValueError):
    """Raised for unparseable versions or constraints."""


@dataclass(frozen=True, order=True)
class Version:
    major: int
    minor: int
    patch: int
    pre: str = ""

    def __str__(self) -> str:
        base = f"{self.major}.{self.minor}.{self.patch}"
        return f"{base}-{self.pre}" if self.pre else base

    @classmethod
    def parse(cls, text: str) -> "Version":
        text = text.strip().lstrip("v")
        m = _VERSION_RE.match(text)
        if not m:
            raise VersionError(f"not a valid version: {text!r}")
        return cls(int(m.group(1)), int(m.group(2)), int(m.group(3)), m.group(4) or "")


@dataclass(frozen=True)
class _Clause:
    op: str
    version: Version

    def matches(self, v: Version) -> bool:
        if self.op in ("==", "="):
            return v == self.version
        if self.op == ">":
            return v > self.version
        if self.op == ">=":
            return v >= self.version
        if self.op == "<":
            return v < self.version
        if self.op == "<=":
            return v <= self.version
        raise VersionError(f"unknown operator: {self.op!r}")


@dataclass(frozen=True)
class Constraint:
    raw: str
    clauses: tuple[_Clause, ...]

    def matches(self, v: Version | str) -> bool:
        if isinstance(v, str):
            v = Version.parse(v)
        return all(c.matches(v) for c in self.clauses)

    @classmethod
    def parse(cls, text: str) -> "Constraint":
        text = text.strip()
        if not text:
            raise VersionError("empty constraint")
        parts = [p.strip() for p in text.split(",") if p.strip()]
        clauses = tuple(_parse_clause(p) for p in parts)
        return cls(raw=text, clauses=clauses)


def _parse_clause(text: str) -> _Clause:
    for op in _OPS:
        if text.startswith(op):
            return _Clause(op=op if op != "=" else "==", version=Version.parse(text[len(op):]))
    # Bare version → exact match.
    return _Clause(op="==", version=Version.parse(text))

File: internal/kernel/plugin/semver.go
package plugin

// Minimal SemVer + constraint parser (no third-party deps).
//
// Mirrors tools/tabula-distro/src/tabula_distro/semver.py — same syntax,
// same semantics. Versions are strict MAJOR.MINOR.PATCH ints with an
// optional -pre.release suffix (compared lexically). Constraints are
// comma-separated AND clauses, e.g. ">=0.8.0,<1.0.0". A bare version
// is treated as exact match. Recognized operators: <, <=, =, ==, >, >=.

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var versionRE = regexp.MustCompile(`^(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?$`)

// Version is a parsed SemVer triple plus optional pre-release tag.
type Version struct {
	Major int
	Minor int
	Patch int
	Pre   string
}

// String renders the canonical "X.Y.Z[-pre]" form.
func (v Version) String() string {
	base := fmt.Sprintf("%d.%d.%d", v.Major, v.Minor, v.Patch)
	if v.Pre != "" {
		return base + "-" + v.Pre
	}
	return base
}

// Compare returns -1, 0, +1 for v < o, ==, >.
func (v Version) Compare(o Version) int {
	if v.Major != o.Major {
		return cmpInt(v.Major, o.Major)
	}
	if v.Minor != o.Minor {
		return cmpInt(v.Minor, o.Minor)
	}
	if v.Patch != o.Patch {
		return cmpInt(v.Patch, o.Patch)
	}
	// Match the Python implementation: pre is compared lexically. This is
	// imprecise per the SemVer spec but adequate for early development; a
	// version without pre is "less" than one with (lex order: "" < "alpha").
	// We don't ship pre-releases yet.
	return strings.Compare(v.Pre, o.Pre)
}

// ParseVersion parses "X.Y.Z" or "vX.Y.Z" (optionally with "-pre" tag).
func ParseVersion(text string) (Version, error) {
	text = strings.TrimSpace(text)
	text = strings.TrimPrefix(text, "v")
	m := versionRE.FindStringSubmatch(text)
	if m == nil {
		return Version{}, fmt.Errorf("not a valid version: %q", text)
	}
	major, _ := strconv.Atoi(m[1])
	minor, _ := strconv.Atoi(m[2])
	patch, _ := strconv.Atoi(m[3])
	return Version{Major: major, Minor: minor, Patch: patch, Pre: m[4]}, nil
}

// Constraint is a parsed comma-AND of clauses, e.g. ">=0.9.0,<1.0.0".
type Constraint struct {
	Raw     string
	Clauses []constraintClause
}

type constraintClause struct {
	op      string
	version Version
}

// Matches reports whether v satisfies the constraint.
func (c Constraint) Matches(v Version) bool {
	for _, cl := range c.Clauses {
		if !cl.matches(v) {
			return false
		}
	}
	return true
}

func (c constraintClause) matches(v Version) bool {
	cmp := v.Compare(c.version)
	switch c.op {
	case "==", "=":
		return cmp == 0
	case ">":
		return cmp > 0
	case ">=":
		return cmp >= 0
	case "<":
		return cmp < 0
	case "<=":
		return cmp <= 0
	}
	return false
}

// ParseConstraint parses a comma-separated AND of clauses. Empty input is
// an error.
func ParseConstraint(text string) (Constraint, error) {
	text = strings.TrimSpace(text)
	if text == "" {
		return Constraint{}, fmt.Errorf("empty constraint")
	}
	var clauses []constraintClause
	for _, part := range strings.Split(text, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		cl, err := parseClause(part)
		if err != nil {
			return Constraint{}, err
		}
		clauses = append(clauses, cl)
	}
	if len(clauses) == 0 {
		return Constraint{}, fmt.Errorf("empty constraint")
	}
	return Constraint{Raw: text, Clauses: clauses}, nil
}

// Order matters: longer ops first to avoid mis-prefixing.
var clauseOps = []string{"<=", ">=", "==", "<", ">", "="}

func parseClause(text string) (constraintClause, error) {
	for _, op := range clauseOps {
		if strings.HasPrefix(text, op) {
			v, err := ParseVersion(text[len(op):])
			if err != nil {
				return constraintClause{}, err
			}
			canonical := op
			if op == "=" {
				canonical = "=="
			}
			return constraintClause{op: canonical, version: v}, nil
		}
	}
	// Bare version → exact match (mirrors Python parser).
	v, err := ParseVersion(text)
	if err != nil {
		return constraintClause{}, err
	}
	return constraintClause{op: "==", version: v}, nil
}

func cmpInt(a, b int) int {
	switch {
	case a < b:
		return -1
	case a > b:
		return 1
	default:
		return 0
	}
}
